tests/test_autoray.py

```diff
@@ -395,35 +435,106 @@ def test_pad(backend):
             res = fn(*args, **kwargs)
             return res + 1
 
         return new_fn
 
     # then check we can wrap the old (previous) function
-    ar.register_function(backend, 'test_register', wrap_fn, wrap=True)
-    assert ar.do('test_register', x) == 2
+    ar.register_function(backend, "test_register", wrap_fn, wrap=True)
+    assert ar.do("test_register", x) == 2
+
+
+@pytest.mark.parametrize("backend", BACKENDS)
+def test_take(backend):
+    if backend == "sparse":
+        pytest.xfail("sparse doesn't support take yet")
+    num_inds = 4
+    A = gen_rand((2, 3, 4), backend)
+    if backend == "jax":  # gen_rand doesn't work with ints for JAX
+        ind = gen_rand((num_inds,), "numpy", dtype="int64")
+    else:
+        ind = gen_rand((num_inds,), backend, dtype="int64")
+
+    # Take along axis 1, and check if result makes sense
+    B = ar.do("take", A, ind, axis=1)
+    assert B.shape == (2, 4, 4)
+    for i in range(num_inds):
+        assert ar.do(
+            "allclose", ar.to_numpy(A[:, ind[0], :]), ar.to_numpy(B[:, 0, :])
+        )
+    assert ar.infer_backend(A) == ar.infer_backend(B)
+
+
+@pytest.mark.parametrize("backend", BACKENDS)
+def test_concatenate(backend):
+    mats = [gen_rand((2, 3, 4), backend) for _ in range(3)]
+
+    # Concatenate along axis 1, check if shape is correct
+    # also check if automatically inferring backend works
+    mats_concat1 = ar.do("concatenate", mats, axis=1)
+    mats_concat2 = ar.do("concatenate", mats, axis=1, like=backend)
+    assert mats_concat1.shape == mats_concat2.shape == (2, 9, 4)
+    assert (
+        backend
+        == ar.infer_backend(mats_concat1)
+        == ar.infer_backend(mats_concat2)
+    )
+
+
+@pytest.mark.parametrize("backend", BACKENDS)
+def test_stack(backend):
+    mats = [gen_rand((2, 3, 4), backend) for _ in range(3)]
+
+    # stack, creating a new axis (at position 0)
+    # also check if automatically inferring backend works
+    mats_stack1 = ar.do("stack", mats)
+    mats_stack2 = ar.do("stack", mats, like=backend)
+    assert mats_stack1.shape == mats_stack2.shape == (3, 2, 3, 4)
+    assert (
+        backend
+        == ar.infer_backend(mats_stack1)
+        == ar.infer_backend(mats_stack2)
+    )
+
+
+@pytest.mark.parametrize("backend", BACKENDS)
+def test_einsum(backend):
+    if backend == "sparse":
+        pytest.xfail("sparse doesn't support einsum yet")
+    A = gen_rand((2, 3, 4), backend)
+    B = gen_rand((3, 4, 2), backend)
+    C1 = ar.do("einsum", "ijk,jkl->il", A, B, like=backend)
+    C2 = ar.do("einsum", "ijk,jkl->il", A, B)
+    C3 = ar.do("reshape", A, (2, 12)) @ ar.do("reshape", B, (12, 2))
+    assert C1.shape == C2.shape == (2, 2)
+    assert ar.do("allclose", ar.to_numpy(C1), ar.to_numpy(C3))
+    assert ar.do("allclose", ar.to_numpy(C2), ar.to_numpy(C3))
+    assert (
+        ar.infer_backend(C1)
+        == ar.infer_backend(C2)
+        == ar.infer_backend(C3)
+        == backend
+    )
```

Comment:
Is this a black preference (no final newline)?

Reply:
No, that's an honest mistake. Black is an autoformatter to be used in conjunction with the `pycodestyle` linter, which also wants a newline at end of file. Most of the time `pycodestyle` and the PEP8 guidelines agree. Biggest difference is the standard maximum line width, which is 79 for PEP8. However on a 1080p screen you can easily fit two editors next to each other with line width ~100 chars, so many people consider this line width a bit too strict.